socs/agents/hwp_supervisor/agent.py

```diff
@@ -0,0 +1,773 @@
+
+        self.success = success
+        self.msg = msg
+        self.state_type = state_type
+
+    def run_and_validate(self, op, kwargs=None, timeout=-1):
+        """
+        Runs an OCS Operation, and validates that it was successful.
+
+        Args
+        -------
+        op : OCS MatchedOp
+            Operation to run. This must be a MatchedOp, or a method of an
+            OCSClient
+        kwargs : dict, optional
+            Kwargs to pass to the operation
+        timeout : float, optional
+            Timeout for the wait command. This defaults to
+            ``default_wait_timeout`` which is 10 seconds. If this is set to
+            None, will wait indefinitely.
+        """
+        if kwargs is None:
+            kwargs = {}
+        if timeout is -1:
+            timeout = self.default_timeout
+        op.start(**kwargs)
+        status, msg, session = op.wait(timeout=timeout)
+
+    def update(self, clients: HWPClients, hwp_state: HWPState):
+        """
+        Runs control operations and updates the control-state based on the
+        ``hwp_state`` object.
+
+        Args
+        --------
+        clients : HWPClients
+            Clients to use to control the HWP.
+        hwp_state : HWPState
+            Current state of the HWP.
+        """
+        if self.state_type == ControlStateType.NONE:
+            return
+
+        elif self.state_type == ControlStateType.PID_FREQ:
+            # Set the PID and then switch to WAIT_FOR_TARGET_FREQ
+            self.run_and_validate(clients.pid.set_direction,
+                                  kwargs={'direction': self.pid_dir})
+            self.run_and_validate(clients.pid.declare_freq,
+                                  kwargs={'freq': self.target_freq})
+            self.run_and_validate(clients.pid.use_ext)
+            self.run_and_validate(clients.pid.set_on)
+            self.set_state_type(ControlStateType.WAIT_FOR_TARGET_FREQ)
+
+        elif self.state_type == ControlStateType.PMX_OFF:
+            # Turn PMX off and then switch to WAIT_FOR_TARGET_FREQ
+            self.run_and_validate(clients.pid.set_off)
+            self.set_state_type(ControlStateType.WAIT_FOR_TARGET_FREQ)
+
+        elif self.state_type == ControlStateType.WAIT_FOR_TARGET_FREQ:
+            # Check if we are close enough to the target frequency.
+            # This will make sure we remain within the frequency threshold for
+            # ``self.freq_thresh_duration`` seconds before switching to DONE
+            f = hwp_state.pid_current_freq
+            if f is not None:
+                # If within_thresh for freq_thresh_duration, switch to Done
+                if np.abs(f - self.target_freq) < self.freq_thresh:
+                    if self.freq_within_thresh_start is None:
+                        self.freq_within_thresh_start = time.time()
+                    elif time.time() - self.freq_within_thresh_start > self.freq_thresh_duration:
```

Comment:
Question about the behavior of `freq_thresh_duration`
If the hwp did't reach close enough to the `target_freq` with in `freq_thresh_duration`

Does the hwp task end at that time? and is the task to be considered as ` success=True`?
I think the warning like "hwp didn't reach `target_freq` within `freq_thresh_duration`" will be useful.

Reply:
`freq_thresh_duration` is the amount of time that the freq has to be within the threshold in order for it to be considered a success. I added this so we don't consider it a success if the HWP frequency just passes through the target frequency but doesn't stop.

Right now there isn't a time limit for which we switch to the DONE state with success=False, but that is something we can easily add if we want.